module/os/freebsd/spl/spl_uio.c

```diff
@@ -105,3 +109,254 @@ zfs_uio_fault_move(void *p, size_t n, zfs_uio_rw_t dir, zfs_uio_t *uio)
+
+		ASSERT3P(page, !=, NULL);
+		ASSERT(vm_page_sbusied(page));
+		vm_page_sunbusy(page);
+	}
+}
+
+#else
+
+static void
+zfs_uio_set_pages_to_stable(zfs_uio_t *uio)
+{
+	ASSERT3P(uio->uio_dio.pages, !=, NULL);
+	ASSERT3U(uio->uio_dio.npages, >, 0);
+
+	for (int i = 0; i < uio->uio_dio.npages; i++) {
+		vm_page_t page = uio->uio_dio.pages[i];
+		ASSERT3P(page, !=, NULL);
+
+		MPASS(page == PHYS_TO_VM_PAGE(VM_PAGE_TO_PHYS(page)));
+		vm_page_busy_acquire(page, VM_ALLOC_SBUSY);
+		pmap_remove_write(page);
+	}
+}
+
+static void
+zfs_uio_release_stable_pages(zfs_uio_t *uio)
+{
+	ASSERT3P(uio->uio_dio.pages, !=, NULL);
+	for (int i = 0; i < uio->uio_dio.npages; i++) {
+		vm_page_t page = uio->uio_dio.pages[i];
+
+		ASSERT3P(page, !=, NULL);
+		vm_page_sunbusy(page);
+	}
+}
+
+#endif
+
+/*
+ * If the operation is marked as read, then we are stating the pages will be
+ * written to and must be given write access.
+ */
+static int
+zfs_uio_hold_pages(unsigned long start, size_t len, unsigned long nr_pages,
+    zfs_uio_rw_t rw, vm_page_t *pages)
+{
+	vm_map_t map;
+	vm_prot_t prot;
+	int count;
+
+	map = &curthread->td_proc->p_vmspace->vm_map;
+	ASSERT3S(len, >, 0);
+
+	prot = rw == UIO_READ ? (VM_PROT_READ | VM_PROT_WRITE) : VM_PROT_READ;
+	count = vm_fault_quick_hold_pages(map, start, len, prot, pages,
+	    nr_pages);
+
+	return (count);
+}
+
+static void
+zfs_uio_unhold_pages(vm_page_t *m, int count)
+{
+#if __FreeBSD_version < 1300050
+	for (int i = 0; i < count; i++) {
+		vm_page_t page = m[i];
+		ASSERT3P(page, !=, NULL);
+		vm_page_lock(page);
+		vm_page_unwire_noq(page);
+		vm_page_unlock(page);
+	}
+#else
+	vm_page_unhold_pages(m, count);
+#endif
+}
+
+void
+zfs_uio_free_dio_pages(zfs_uio_t *uio, zfs_uio_rw_t rw)
+{
+	if (!(uio->uio_extflg & UIO_DIRECT))
+		return;
+
+	ASSERT3P(uio->uio_dio.pages, !=, NULL);
+	ASSERT(zfs_uio_rw(uio) == rw);
+
+	if (rw == UIO_WRITE)
+		zfs_uio_release_stable_pages(uio);
+
+	zfs_uio_unhold_pages(&uio->uio_dio.pages[0],
+	    uio->uio_dio.npages);
+
+	kmem_free(uio->uio_dio.pages,
+	    uio->uio_dio.npages * sizeof (vm_page_t));
+}
+
+static long
+zfs_uio_get_user_pages(unsigned long start, unsigned long nr_pages,
+    size_t len, zfs_uio_rw_t rw, vm_page_t *pages)
+{
+	int count;
+
+	count = zfs_uio_hold_pages(start, len, nr_pages, rw, pages);
+
+	if (count != nr_pages) {
+		if (count > 0)
+			vm_page_unhold_pages(pages, count);
+		return (count);
+	}
+
+	ASSERT3U(count, ==, nr_pages);
+
+#if __FreeBSD_version < 1300050
+	for (int i = 0; i < count; i++) {
+		vm_page_t page = pages[i];
+		vm_page_lock(page);
+		vm_page_wire(page);
+		vm_page_unhold(page);
+		vm_page_unlock(page);
+	}
+#endif
+
+	return (count);
+}
+
+static size_t
+zfs_uio_iov_step(struct iovec v, zfs_uio_t *uio, int *numpages)
+{
+	unsigned long addr = (unsigned long)(v.iov_base);
+	size_t len = v.iov_len;
+	int n = DIV_ROUND_UP(len, PAGE_SIZE);
+
+	int res = zfs_uio_get_user_pages(P2ALIGN(addr, PAGE_SIZE - 1), n, len,
```

Comment:
I can only imagine you're going for a page aligned address:
```suggestion
	int res = zfs_uio_get_user_pages(P2ALIGN(addr, PAGE_SIZE), n, len,
```

Reply:
Good catch! Yes, this was a mistake on my part. I have update this.